examples/animatediff/ad/modules/diffusionmodules/motion_module.py

```diff
@@ -0,0 +1,523 @@
+            return x
+
+        def _rearrange_out(x, d=None):
+            #  (b*d f c) -> (b d f c) -> (b f d c) -> (b*f d c)
+            bd, f, c = x.shape
+            b = bd // d
+            x = x.reshape(b, d, f, c).permute(0, 2, 1, 3).reshape(b * f, d, c)
+            return x
+
+        d = hidden_states.shape[1]
+        # hidden_states = rearrange(hidden_states, "(b f) d c -> (b d) f c", f=video_length)
+        hidden_states = _rearrange_in(hidden_states, f=video_length)
+
+        if self.pos_encoder is not None:
+            hidden_states = self.pos_encoder(hidden_states)
+
+        # encoder_hidden_states = encoder_hidden_states
+        # encoder_hidden_states = encoder_hidden_states
+
+        if self.group_norm is not None:
+            hidden_states = self.group_norm(hidden_states.transpose(1, 2)).transpose(1, 2)
+
+        query = self.to_q(hidden_states)
+        # dim = query.shape[-1]
+
+        # (b*d f attn_dim*num_heads) -> (b*d*num_heads f attn_dim), where c=attn_dim*num_heads
+        query = self.reshape_heads_to_batch_dim(query)
+
+        if self.added_kv_proj_dim is not None:
+            raise NotImplementedError
+
+        encoder_hidden_states = encoder_hidden_states if encoder_hidden_states is not None else hidden_states
+        key = self.to_k(encoder_hidden_states)
+        value = self.to_v(encoder_hidden_states)
+
+        key = self.reshape_heads_to_batch_dim(key)
+        value = self.reshape_heads_to_batch_dim(value)
+
+        if attention_mask is not None:
+            if attention_mask.shape[-1] != query.shape[1]:
+                target_length = query.shape[1]
+                # TODO:
+                attention_mask = ops.pad(attention_mask, (0, target_length), value=0.0)
+                attention_mask = attention_mask.repeat_interleave(self.heads, dim=0)
+
+        # attention, what we cannot get enough of
+        hidden_states = self._attention(query, key, value, attention_mask)
+
+        # linear proj
+        # TODO: merge with dropout using Sequential?
+        hidden_states = self.to_out(hidden_states)
+
+        # dropout
+        # hidden_states = self.to_out[1](hidden_states)
+
+        # hidden_states = rearrange(hidden_states, "(b d) f c -> (b f) d c", d=d)
+        hidden_states = _rearrange_out(hidden_states, d=d)
+
+        return hidden_states
+
+    def _attention(self, query, key, value, attention_mask=None):
+        """
+        Args:
+            q, k, v shape: (b*d*num_heads f attn_dim)
+        Returns:
+            out: (b*d f attn_dim*num_heads) = (b*d f c)
+        """
+        if self.upcast_attention:
+            query = query.to_float(ms.float32)
+            key = key.to_float(ms.float32)
+
+        attention_scores = ms.ops.matmul(query, ms.ops.transpose(key, (0, 2, 1))) * self.scale
+
+        if attention_mask is not None:
+            attention_scores = attention_scores + attention_mask
+
+        if self.upcast_softmax:
```

Comment:
Why not casting `attention_scores` to fp32 before adding it with `attention_mask`

Reply:
should be fine. the precision is tested, mre < 10e-5